contributing guidelines (#381)
smartcontractkit/documentation: docs/Developer Reference/contributing-to-chainlink.md

---
layout: nodes.liquid
section: smartContract
date: Last Modified
title: "Contributing to Chainlink"
permalink: "docs/contributing-to-chainlink/"
---
Chainlink is an open-source project licensed [under the MIT license](https://github.com/smartcontractkit/chainlink/blob/master/LICENSE), and we encourage contributions from all developers and community members.

[insert YT video here]

# What It Means to Contribute
When you contribute to the Chainlink project, you as a developer or community member contribute your time and effort to help improve and grow Chainlink. Your contribution can be from various methods:

> - Building and maintaining the Chainlink software and tools
> - Improving and maintaining the documentation, including translations into other languages
> - Creating Chainlink focused content (blog posts, tutorials, videos etc)
> - Becoming a developer expert
> - Running an in-person meetup or watch party
> - Running a Chainlink focused developer Bootcamp (in person or online)
> - Participate in a hackathon
> - Applying for a grant

Reviewer: ```suggestion
-[ Building and maintaining the Chainlink software and tools](https://docs.chain.link/docs/contributing-to-chainlink/#contributing-to-software-and-tooling)
- [Improving and maintaining the documentation, including translations into other languages](https://docs.chain.link/docs/contributing-to-chainlink/#contributing-to-the-documentation)
- [Creating Chainlink focused content (blog posts, tutorials, videos etc)](https://docs.chain.link/docs/contributing-to-chainlink/#creating-community-content)
- [Becoming a developer expert](https://docs.chain.link/docs/contributing-to-chainlink/#becoming-a-developer-expert)
- [Becoming a community advocate](https://docs.chain.link/docs/contributing-to-chainlink/#joining-the-chainlink-community-advocate-program)
- [Running a Chainlink focused developer Bootcamp (in person or online)](https://docs.chain.link/docs/contributing-to-chainlink/#running-a-chainlink-focused-developer-bootcamp)
- [Running an in-person meetup or watch party](https://docs.chain.link/docs/contributing-to-chainlink/#running-an-in-person-meetup-or-watch-party)
- [Participate in a hackathon](https://docs.chain.link/docs/contributing-to-chainlink/#participate-in-a-hackathon)
- [Applying for a grant](https://docs.chain.link/docs/contributing-to-chainlink/#applying-for-a-grant)
```

Reviewer: Added links to the "What it means to contribute" section that sends the user to the appropriate anchor tags on the page, so they can go exactly where they want without searching the page. Also, re-ordered some items and added "becoming a community advocate" to match the order and sections on the page.

Author: thanks @ZakAyesh . I added your suggestions, but changed the URLs to not be absolute, ie just the anchors, so the links also work while browsing the preview site

#### Voting on new features

Sometimes a new feature is put to a vote to decide if it's something that the team and wider community should implement. When an feature is put to a vote, the issue is tagged with the 'needs votes' label. You can contribute to the voting process by reacting to the first post in the feature request with a thumbs up or thumbs down emoji. This will help drive the decision. You can also contribute your thoughts by replying directly to the feature request with a new post in the thread.

![Voting on a new feature request](/files/voting-issue.png)

### Submitting a Pull Request

The best way to contribute to Chainlink is to submit a [pull request (PR)](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests). PRs can be submitted for various reasons, such as fixing an identified issue, adding a feature or improvement to the project, or even adding an entirely new repository to the Chainlink source code for a new tool or feature.

> If you're looking for something to pick up and create a PR for, you can search through the Chainlink repositories to find open issues, and approved feature requests. Alternatively, you can take a look at the current priorities for development listed on the Chainlink [Pivotal Tracker](https://www.pivotaltracker.com/n/projects/2129823).

Reviewer 2: We don't actually use Pivotal anymore, and I don't think we publish the new alternative.

Author: thanks @alexroan. Removed that sentence

---
layout: nodes.liquid
section: smartContract
date: Last Modified
title: "Contributing to Chainlink"
permalink: "docs/contributing-to-chainlink/"
---
Chainlink is an open-source project licensed [under the MIT license](https://github.com/smartcontractkit/chainlink/blob/master/LICENSE), and we encourage contributions from all developers and community members.

[insert YT video here]

# What It Means to Contribute

> To contribute to the Chainlink project means you as a developer or community member contribute your time and effort to help improve and grow Chainlink via various methods such as:

Reviewer 3: ```suggestion
When you contribute to the Chainlink project, you as a developer or community member contribute your time and effort to help improve and grow Chainlink. Your contribution can be from various methods:
```

> - Build and maintain the Chainlink software and tools

Reviewer 3: ```suggestion
- Building and maintaining the Chainlink software and tools
```

Parallel with the other bullet points

- Improving and maintaining the documentation, including translations into other languages
- Creating Chainlink focused content (blog posts, tutorials, videos etc)
- Becoming a developer expert
- Running an in-person meetup or watch party
- Running a Chainlink focused developer Bootcamp (in person or online)
- Participate in a hackathon
- Applying for a grant

# Why Should You Contribute

> [Open source software](https://en.wikipedia.org/wiki/Open-source_software) is a model that brings multiple benefits for both the project as well as the contributors. As a developer or community member, contributing to Chainlink helps you to gain valuable skills and experience, improve the software you and many others rely on and use, as well as grow your personal brand in the community which can lead to future employment opportunities. On top of these awesome things, contributing to open source is fun,  gives you a real sense of community involvement, and gives you a real personal sense of satisfaction knowing that you're part of something bigger, that's helping to build something that will enable a fairer, more transparent and efficient new world.

Reviewer 3: ```suggestion
[Open source software](https://en.wikipedia.org/wiki/Open-source_software) is a model that brings multiple benefits for both the project and the contributors. As a developer or community member, contributing to Chainlink helps you to gain valuable skills and experience, improve the software that you use, and grow your personal brand in the community which can lead to future employment opportunities. On top of these awesome things, contributing to open source is fun.  It can give you a sense of community involvement, and gives you a personal sense of satisfaction knowing that you're part of an effort to build something that will enable a fairer, more transparent, and efficient new world.
```

When you first create an issue, you will be asked to categorize it. This will then prefix the issue name to give viewers an indication of what category the issue relates to:
- [NODE]: The issue relates to the core node software
- [DEVEL]: The issue is a result of working on code found in the current repository
- [FEAT]: The issue relates to a new feature request
- [SMRT]: The issue related to using Chainlink smart contracts
- [EXPL]: The issue related to using the Chainlink Explorer
- [FAUC]: The issue related to using the Chainlink Faucet

![Selecting the new issue category](/files/new-issue-category.png)

Once you select a category, you will be given the opportunity to enter the details for the issue. Please include as much detail as possible. This includes a thorough description, environment and software version details, and detailed steps on how to reproduce the issue. The more thorough you are here, the better the chances someone will be able to pick up the issue and resolve it.

> Once the issue has been successfully created, someone should respond within a day or two and acknowledge the issue has been received, and they will then tag the issue with an appropriate label. You should then monitor the state of the open issue for any questions or updates.

Author: we try to respond within a day or two, but we don't have a firm SLA as far as I'm aware. Will update

All code changes should try to follow the [style guide] (https://github.com/smartcontractkit/chainlink/wiki/Code-Style-Guide), and all PRs should be in a branch appropriately named (ie feat/feature-description, or devel/issue-description) Once you submit a PR, you should get a response by a team member within a day or two acknowledging that the PR has been received. After that, be sure to monitor the PR for any additional questions or updates that come up while the PR gets reviewed by the team and the wider community.

## Contributing to the documentation
The [Chainlink documentation](https://docs.chain.link/) is the go-to place for developers wanting to learn how to use Chainlink. The documentation is [open source](https://github.com/smartcontractkit/documentation), allowing for other developers and community members to contribute to adding or improving it. You can contribute to the Chainlink documentation in various ways:
- Improving the readability of pages
- Fixing typos or grammar errors
- Adding new sections you feel would be useful

> - Translating the documentation into other languages.

Reviewer 3: ```suggestion
- Translating the documentation into other languages
```

> The process for contributing to the documentation follows the process defined earlier in the [Submitting a Pull Request](#submitting-a-pull-request) section. Each page also has a 'Suggest Edits' link on the top right, and will directly take you to the page in the [documentation repository](https://github.com/smartcontractkit/documentation), where you can then create a new PR with the suggested changes. If creating a PR for the documentation, please read the [contributing guidelines](https://github.com/smartcontractkit/documentation/blob/main/CONTRIBUTING.md).

Reviewer 3: ```suggestion
The process for contributing to the documentation follows the process defined earlier in the [Submitting a Pull Request](#submitting-a-pull-request) section. Each page also has a 'Suggest Edits' link on the top right, and will directly take you to the page in the [documentation repository](https://github.com/smartcontractkit/documentation), where you can create a new PR with the suggested changes. Before you create a PR for the documentation, read the [contributing guidelines](https://github.com/smartcontractkit/documentation/blob/main/CONTRIBUTING.md).
```

> If you wish to translate the documentation into a new language not yet supported, feel free to [reach out to the team](mailto:developer-relations@chainlinklabs.com) beforehand, so we can make sure you get the support you need.

Reviewer 3: ```suggestion
If you want to translate the documentation into a new language that is not yet supported, feel free to [reach out to the team](mailto:developer-relations@chainlinklabs.com) beforehand, so we can make sure you get the support you need.
```

## Creating Community Content

> Chainlink has a strong, vibrant community of developers and community advocates. Community members often create Chainlink-focused content in various forms and have them published for the wider community on various platforms as a way to increase knowledge and awareness of Chainlink solutions across the wider community, as well as building up their own personal skills and brand in the community.

Reviewer 3: ```suggestion
Chainlink has a strong and vibrant community of developers and community advocates. Community members often create Chainlink-focused content in various forms and publish it for the wider community on various platforms. This increases knowledge and awareness of Chainlink solutions across the wider community and builds the contributor's personal skills and brand in the community.
```

Some examples of the content generated from the community:
- Document your experience in using Chainlink as part of your project
- Do a deep dive blog post or video on a Chainlink solution
- Write up technical tutorials showcasing Chainlink being used in various use cases

## Becoming a Developer Expert

> Chainlink Developer Experts are smart contract and blockchain developers with deep experience building with Chainlink who are passionate about sharing their technical knowledge with the world. As a developer expert, you will receive recognition in the community, previews of new Chainlink features, exclusive access to Chainlink events, and opportunities to level up your technical and soft skills. You can apply to become a developer expert via the [Chainlink Developer Experts page](https://chain.link/developers/experts).

Reviewer 3: ```suggestion
Chainlink Developer Experts are smart contract and blockchain developers with deep experience building applications using Chainlink. They are passionate about sharing their technical knowledge with the world. As a developer expert, you will receive recognition in the community, previews of new Chainlink features, exclusive access to Chainlink events, and opportunities to level up your technical and soft skills. You can apply to become a developer expert on the [Chainlink Developer Experts page](https://chain.link/developers/experts).
```

## Joining the Chainlink Community Advocate program

> The [Chainlink Community Advocate Program](https://blog.chain.link/expanding-the-chainlink-community-advocate-program/) is a program designed to help accelerate the awareness and adoption of Chainlink. Chainlink community advocates are passionate members of the Chainlink community that help to achieve this by running virtual and in-person meetups, connecting with partners and sponsors, creating various content, work directly with the teams that are making Chainlink-powered smart contracts, as well as other tasks related to growing awareness and adoption of Chainlink. Many Advocates have gone on to have rewarding careers in the blockchain industry and working on Chainlink specifically.

Reviewer 3: ```suggestion
The [Chainlink Community Advocate Program](https://blog.chain.link/expanding-the-chainlink-community-advocate-program/) is a program designed to help accelerate the awareness and adoption of Chainlink. Chainlink community advocates are passionate members of the Chainlink community that help to achieve this by running virtual and in-person meetups, connecting with partners and sponsors, creating content, and working directly with the teams that are making Chainlink-powered smart contracts. Many Advocates have gone on to have rewarding careers in the blockchain industry, and some of them work on Chainlink specifically.
```

To become a community advocate, you can apply via the [community advocates web page](https://chain.link/community/advocates).

## Running a Chainlink Focused Developer Bootcamp

> In June 2021, the very first [Chainlink Developer Bootcamp](http://chain.link/bootcamp) was [virtually hosted](https://blog.chain.link/smart-contract-developer-bootcamp-on-demand/) by the Chainlink Labs team. If you're passionate about educating others on smart contracts and Chainlink, you can contribute by running your own virtual or in-person developer Bootcamp in your city. You can also contribute by translating and running the developer Bootcamp in another language. Before running your own Bootcamp, we encourage you to first [reach out to the team](mailto:developer-relations@chainlinklabs.com) so we can make sure you have the support you need.

Reviewer 3: ```suggestion
In June 2021, the Chainlink Labs team [virtually hosted](https://blog.chain.link/smart-contract-developer-bootcamp-on-demand/) the first [Chainlink Developer Bootcamp](http://chain.link/bootcamp). If you're passionate about educating others about smart contracts and Chainlink, you can contribute by running your own developer Bootcamp. You can also contribute by translating an existing Bootcamp and running it in another language. Before you run your own Bootcamp, [reach out to the team](mailto:developer-relations@chainlinklabs.com) so we can make sure you have the support that you need.
```

## Running an In-Person Meetup or Watch Party
If you're passionate about helping to grow the awareness and adoption of Chainlink, you can contribute by running an in-person meetup or watch party for a Chainlink event such as [SmartCon](https://www.smartcontractsummit.io/). Meetups are a great way to meet others also passionate about how hybrid smart contracts can create an economically fair world.

> If you're interested in running an in-person meetup or watch party, feel free to [reach out to the team](mailto:community@smartcontract.com) so we can make sure you have the support you need.

Reviewer 3: ```suggestion
If you're interested in running an in-person meetup or watch party, [reach out to the team](mailto:community@smartcontract.com) so we can make sure you have the support that you need.
```

## Participate in a Hackathon

> Chainlink runs hackathons multiple times per year and often sponsors other hackathons across the blockchain ecosystem. Participating in a hackathon that Chainlink is a part of is a great way for learning how to use Chainlink, and also a great way to showcase your skills to the Chainlink team as well as the wider community. Hackathons are also a very popular place for recruiting talent into the blockchain ecosystem.

Reviewer 3: ```suggestion
Chainlink runs hackathons multiple times per year and often sponsors other hackathons across the blockchain ecosystem. Participating in a hackathon that Chainlink is a part of is a great way to learn how to use Chainlink. It is also a great way to showcase your skills to the Chainlink team and the wider community. Hackathons are a popular place for recruiting talent into the blockchain ecosystem.
```

> To stay up to date on the hackathons Chainlink is running or sponsoring, keep an eye out on official Chainlink social media channels, and sign up for our [developer newsletter](https://docs.chain.link/docs/developer-communications/).

Reviewer 3: ```suggestion
To stay up to date on the hackathons that Chainlink is running or sponsoring, keep an eye out on the official Chainlink social media channels, and sign up for our [developer newsletter](https://docs.chain.link/docs/developer-communications/).
```

Add a link for "official Chainlink social media channels" as well?

## Applying for a Grant

> The [Chainlink grant program](https://chain.link/community/grants) encourages the community to focus on the creation of critical developer tooling, the addition of high-quality data, and the launching of key services around the Chainlink Network. Grant categories include community, integration, bug bounty, research, and social impact grants. If you believe you have a great idea that fits into one of these categories, you can apply for a grant. If successful, you will receive the funding and support needed to successfully build and implement it.

Reviewer 3: ```suggestion
The [Chainlink grant program](https://chain.link/community/grants) encourages the community to create critical developer tooling, add high-quality data, and the launch key services around the Chainlink Network. Grant categories include community, integration, bug bounty, research, and social impact grants. If you have a great idea that fits into one of these categories, you can apply for a grant. If successful, you will receive the funding and support needed to successfully build and implement your idea.
```

> For more information, head to the [Chainlink Grants web page](https://chain.link/community/grants)

Reviewer 3: ```suggestion
For more information about the grant program, go to the [Chainlink Grants web page](https://chain.link/community/grants).
```

> 
> 
> 
> 

Reviewer 3: ```suggestion
```
Only need one newline at the end.
